finish templating VectorTools; introduce templates to MatrixTools and…

davydden opened the pull request:
… DoFTools

1.  drop the typedef FunctionMap from (hp::)DoFHandler and template n_boundary_dofs().
 Fix a few places where it was used.
2.  template make_boundary_sparsity_pattern(...,FunctionMap,...)
3.  number templates in MatrixCreator::
  create_mass_matrix(), create_boundary_mass_matrix(),
  apply_boundary_values() and local_apply_boundary_values().
4.  number templates in VectorTools::
  interpolate_boundary_values(), project_boundary_values(),
5.  instantiate ConstraintMatrix::distribute_local_to_global for SparseMatrix(long double)
6.  Adjust a few unit tests to explicitly state templated arguments.
Unfortunately this has to be done due to the usage of nested types
for which one can’t infer template arguments.

fixes https://github.com/dealii/dealii/issues/2236 ;
in reference to https://github.com/dealii/dealii/issues/2033 ;

Sorry for the big PR, almost everything is related :confused: .
Unit tests in `bits/`, `numerics/`, `vector_tools/` pass for me in full on `clang`.

Rombur commented:
Just a general question. I see that you modified step-51 in the tests but not in the examples. Did you check that all the tutorials still work?

davydden commented:
correct, has to be changed as well. Will do that after we are settled on the content.

bangerth commented on include/deal.II/dofs/dof_handler.h:
```diff
@@ -1098,7 +1094,8 @@ class DoFHandler  :  public Subscriptor
 template <> types::global_dof_index DoFHandler<1>::n_boundary_dofs () const;
-template <> types::global_dof_index DoFHandler<1>::n_boundary_dofs (const FunctionMap &) const;
+//template <> template <> types::global_dof_index DoFHandler<1>::n_boundary_dofs<double> (const FunctionMap<double> &) const;
+//template <> template <> types::global_dof_index DoFHandler<1>::n_boundary_dofs<float> (const FunctionMap<float> &) const;
```
If you don't need these, just remove them.

davydden replied:
i am not sure if these are needed or not. If i understand it correctly, it's a promise that there will be an explicit specializations for those cases. But if I commeit it out while keeping the specialized versions in .cc file it must be all good, right?

bangerth commented on include/deal.II/dofs/dof_handler.h:
```diff
@@ -707,8 +702,9 @@ class DoFHandler  :  public Subscriptor
+  template<typename number>
   types::global_dof_index
-  n_boundary_dofs (const FunctionMap &boundary_ids) const;
+  n_boundary_dofs (const typename dealii::FunctionMap<spacedim,number>::type &boundary_ids) const;
```
This doesn't work. `number` is not deducible in the current context, so you force every user to explicitly specify the template argument. This is not backward compatible, nor convenient.

Just remove the `FunctionMap` altogether and use the underlying `::type` typedef. You may want to say in the documentation of the class that the type of `boundary_ids` equals `typename FunctionMap<spacedim,number>::type`.

bangerth commented on include/deal.II/numerics/matrix_tools.h:
```diff
@@ -53,6 +53,7 @@ namespace hp
 
 #ifdef DEAL_II_WITH_PETSC
+#  include <petscvec.h>
 namespace PETScWrappers
```
add an empty line here.

davydden replied:
after or before include?

bangerth commented on include/deal.II/numerics/matrix_tools.h:
```diff
@@ -68,6 +69,8 @@ namespace PETScWrappers
 #ifdef DEAL_II_WITH_TRILINOS
+// TODO: move this to a stand-alone header and include from there?
+#  define TrilinosScalar double
 namespace TrilinosWrappers
```
Add empty line.

I'm not in favor of this here, since it adds a global typedef. Why is this necessary?

davydden replied:
because we forward declare all the trilinos-related objects and I want to define functions with consistent data types.

`TrilinosScalar` seems to be an invention of deal.ii, it is defined in (including this)
```
$ ag --cpp "define TrilinosScalar"
include/deal.II/lac/trilinos_block_sparse_matrix.h
34:#  define TrilinosScalar double

include/deal.II/lac/trilinos_vector_base.h
35:#  define TrilinosScalar double

include/deal.II/lac/trilinos_sparse_matrix.h
36:#  define TrilinosScalar double

include/deal.II/numerics/matrix_tools.h
75:#  define TrilinosScalar double
```
I would say this definition should be moved to a stand-alone header like `lac/trilinos_types.h` and included where needed. What do you think?

Rombur replied:
Don't bother with ```TrilinosScalar```, I am pretty sure that in some files we use double directly. Epetra only supports double so it doesn't matter anyway. When we will add Tpetra, we can add a typedef TrilinosScalar in ```base/types.h```.

davydden replied:
then i would prefer to keep this extra `define` here so that all functions are consistently defined. Extra define will be cleaned up together with others when Tpetra is here.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -698,26 +698,26 @@ namespace VectorTools
    * See the general documentation of this namespace for more information.
    */
-  template <typename DoFHandlerType>
+  template <typename number, typename DoFHandlerType>
```
here and below: you make an incompatible change by adding the new template argument as first argument. This breaks all code where someone already specifies the dof handler type explicitly. better to add the new template argument last.

bangerth commented on include/deal.II/numerics/vector_tools.h:
```diff
@@ -698,26 +698,26 @@ namespace VectorTools
    * hp::DoFHandler object.
    */
-  template <int dim, int spacedim>
+  template <int dim, typename number, int spacedim>
```
same here. there's also no good reason to separate `dim` and `spacedim` :-)

bangerth commented on tests/bits/step-51p.cc:
```diff
@@ -343,10 +343,10 @@ namespace Step51
-                                          QGauss<dim-1>(fe.degree+1),
-                                          constraints);
+    VectorTools::project_boundary_values<dim,double> (dof_handler,
+                                                      boundary_functions,
```
the reason why you need to specify the `number` type here and elsewhere is because of the use of `FunctionMap::type` which makes `number` non-deducible for a compiler. You can avoid all of these changes to the testsuite if you make `number` deducible to the compiler.

bangerth commented on source/numerics/vector_tools_boundary.inst.in:
```diff
@@ -19,15 +19,15 @@ for (deal_II_dimension : DIMENSIONS; deal_II_space_dimension : SPACE_DIMENSIONS;
     namespace VectorTools \{
       template
-        void interpolate_boundary_values
+        void interpolate_boundary_values<double>
```
another example of a non-deducible context. the changes in this file are also all avoidable if you make `number` deducible.

bangerth commented:
This patch demonstrates why `FunctionMap` was a bad idea all along. You have my blessing to replace all occurrences of `typename FunctionMap<dim,number>::type` by `std::map<...,....>` right away. This allows the compiler to deduce `dim` and `number`, which it cannot otherwise.

bangerth commented:
(This would also have saved you a lot of work in changing tests and instantiations :-( )

davydden commented:
will do that

davydden commented:
btw, thanks @bangerth for pointing to the usage of `map<>` directly. I am glad that we can keep type deduction.

davydden commented:
@bangerth i addressed your comments (except for `TrilinosScalar` for now). Ready for re-review. Unit tests in `bits/`, `numerics/`, `vector_tools/` pass for me in full on clang.

tjhei commented:
Why do you need ``TrilinosScalar``? Isn't that always going to be a double? If you decide to keep it, can't you use a ``typedef`` instead of a define?

davydden commented:
@tjhei i don't want to hardcode anything, even if it is always `double` for this particular linear algebra. It should be nice and clean, the same way it is for `PetscScalar` and PETSc matrices/vectors. 

Whether it is a typedef or define is not important here as it is already defined in a number of places:
```
$ ag --cpp "define TrilinosScalar"
include/deal.II/lac/trilinos_block_sparse_matrix.h
34:#  define TrilinosScalar double

include/deal.II/lac/trilinos_vector_base.h
35:#  define TrilinosScalar double

include/deal.II/lac/trilinos_sparse_matrix.h
36:#  define TrilinosScalar double

include/deal.II/numerics/matrix_tools.h
75:#  define TrilinosScalar double
```
If you want I can remove all those define's and typedef it inside `base/types.h` (but not in a Trilinos namespace, so that it can be used straight-away the same way PetscScalar is used). Then it will be one place where `TrilinosScalar` is introduced. And if some day it will become `long double` or something, there will be one place to change it.

Rombur commented:
```Epetra``` is not templated so it can never change. But since ```TrilinosScalar``` has been introduced, I guess it would be better to define only in one place.

davydden commented:
@tjhei @Rombur done.

Rombur commented:
It's good for me but I will let @tjhei and @bangerth some time to see the patch before merging.

davydden commented:
@Rombur sure, no rush. I bet @bangerth would like to have a second look.

tjhei commented:
> Whether it is a typedef or define is not important here as it is already defined in a number of places:

Interesting, I didn't know that. I would prefer a typedef over a define if this doesn't break anything.

davydden commented:
> Interesting, I didn't know that. I would prefer a typedef over a define if this doesn't break anything.
no, it does not. I changed to a single `typedef`.

bangerth commented on include/deal.II/base/types.h:
```diff
@@ -130,6 +130,10 @@ namespace types
+ * Declare type used in Epetra.
+ */
+typedef double TrilinosScalar;
 namespace TrilinosWrappers
```
Let's add one or two empty lines between the typedef and the namespace, just to make it clear that they are entirely separate concerns.

In an ideal world, I would have moved the typedef inside the namespace, but that may or may not work any more since there was the previous `#define`...

bangerth commented on include/deal.II/numerics/matrix_tools.h:
```diff
@@ -53,6 +53,8 @@ namespace hp
 
 
 #ifdef DEAL_II_WITH_PETSC
+#  include <petscvec.h>
```
Do you really need this here, in the header file?

bangerth replied:
Ah, I see where you use it further down below. But shouldn't it be enough to `#include` the `petscmath.h` file?

davydden replied:
i get compiler errors when i do this
```
In file included from /Users/davydden/libs-sources/deal.ii/davydden/source/multigrid/mg_tools.cc:36:
In file included from /Users/davydden/libs-sources/deal.ii/davydden/include/deal.II/numerics/matrix_tools.h:56:
/usr/local/include/petscmath.h:278:9: error: unknown type name 'PetscReal'
typedef PetscReal PetscScalar;
        ^
/usr/local/include/petscmath.h:283:1: error: unknown type name 'PETSC_STATIC_INLINE'
PETSC_STATIC_INLINE PetscReal PetscAbsScalar(PetscScalar a) {return a < 0.0 ? -a : a;}
^
...
```

davydden replied:
it's `petscsys.h `. I will change it.

bangerth commented on include/deal.II/numerics/matrix_tools.h:
```diff
@@ -232,23 +234,23 @@ namespace MatrixCreator
                            const Quadrature<dim>    &q,
                            SparseMatrix<number>     &matrix,
-                           const Function<spacedim> *const a = 0,
+                           const Function<spacedim,number> *const a = 0,
```
Here and in many of the functions below, I don't think it's actually necessary that the `Function` argument uses the same scalar type as what we use in the matrix and vector arguments to the function. It's certainly not wrong, and I don't in principle object to it, but it *is* an incompatible change to require this. I just wanted to put it up for discussion.

davydden replied:
that's correct that one could generally use, say, `SparseMatrix<float>` with `Function<dim,double>`. However i think we already made this choice when we reworked `FEValues` last spring https://github.com/dealii/dealii/pull/918 where, for example, the second `number` template argument of `get_function_values` was removed in favour of `typename InputVector::value_type`. 

I see two options here: either to ask it to be consistent (which is the same to what we already did) or to introduce extra template parameter (which is IMHO an overkill).

I guess in 99% it is going to be used with `double` everywhere.

davydden replied:
Until this PR there was an instantiation of `create_mass_matrix` for `SparseMatrix<float>` and `Function<double>`. But this certainly was not used wthin deal.ii. Funny enogh `create_boundary_mass_matrix` was instantiated only for double-double.

bangerth replied:
Yes, it's these kinds of inconsistencies that made it tricky :-( I agree that few people likely used it.

The comparison with FEValues is not entirely accurate. There, we made sure that the output argument had a type that is consistent with all inputs. In some sense, we just made sure that the output has a type that can accommodate the types of all inputs. Here, however, we often have multiple inputs (including, for example, the type of shape functions, coefficients, etc). In principle, we should probably templatize on *all* input types, and enforce the then correct output type. You are correct that that seems like overkill, so let's not worry about it.

bangerth commented on source/dofs/dof_tools_sparsity.inst.in:
```diff
@@ -13,7 +13,47 @@
 
+    //template void
+    //DoFTools::make_boundary_sparsity_pattern<hp::DoFHandler<deal_II_dimension,deal_II_dimension+1>,SP,scalar>
+    //(const hp::DoFHandler<deal_II_dimension,deal_II_dimension+1>& dof,
```
If you don't need it, just remove it. Or, if you think we may later need it, leave it but add a comment why you comment it out.

davydden replied:
it already was commented out. I actually don't know why. So i thought i wouldn't remove it, but just fix the comment part so that it is consistent with the changes.

bangerth replied:
OK

bangerth commented:
This looks pretty alright to me. Can you
* make the couple of minor changes requested above
* add an entry into `changes.h` in the "Incompatibilities" section stating that many functions in VectorTools and MatrixTools now require matching data types between vectors, matrices, and Function<dim> arguments?

Please merge once done.
